xds/internal/xdsclient/xds.go

```diff
@@ -561,6 +562,33 @@ func routesProtoToSlice(routes []*v3routepb.Route, logger *grpclog.PrefixLogger,
 	return routesRet, nil
 }
 
+func hashPoliciesProtoToSlice(policies []*v3routepb.RouteAction_HashPolicy) []*HashPolicy {
+	var hashPoliciesRet []*HashPolicy
+	for _, p := range policies {
+		var policy HashPolicy
+		policy.Terminal = p.Terminal
+		switch p.GetPolicySpecifier().(type) {
+		case *v3routepb.RouteAction_HashPolicy_Header_:
+			policy.HashPolicyType = HashPolicyTypeHeader
+			policy.HeaderName = p.GetHeader().HeaderName
+			policy.Regex = p.GetHeader().GetRegexRewrite().GetPattern().Regex
+			policy.RegexSubstitution = p.GetHeader().GetRegexRewrite().Substitution
+		// ConnectionProperties in Envoy specifies to hash the source ip
+		// address. In grpc, this is logically equivalent to hashing the
+		// channel pointer.
+		case *v3routepb.RouteAction_HashPolicy_ConnectionProperties_:
```

Comment:
We don't support connection properties I think.

The way to specify channel_id is the filter_state field set to contain the key "io.grpc.channel_id".
http://doc/10J27vUqOYH8du6XZLcnmwOOSwO9r3HjIe_Jybn6AbVE